hooks/hookexecution/execution.go

```diff
@@ -0,0 +1,233 @@
+package hookexecution
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prebid/prebid-server/hooks"
+	"github.com/prebid/prebid-server/hooks/hookstage"
+)
+
+type hookResponse[T any] struct {
+	Err           error
+	ExecutionTime time.Duration
+	HookID        HookID
+	Result        hookstage.HookResult[T]
+}
+
+type hookHandler[H any, P any] func(
+	context.Context,
+	hookstage.ModuleInvocationContext,
+	H,
+	P,
+) (hookstage.HookResult[P], error)
+
+func executeStage[H any, P any](
+	executionCtx executionContext,
+	plan hooks.Plan[H],
+	payload P,
+	hookHandler hookHandler[H, P],
+) (StageOutcome, P, stageModuleContext, *RejectError) {
+	stageOutcome := StageOutcome{}
+	stageOutcome.Groups = make([]GroupOutcome, 0, len(plan))
+	stageModuleCtx := stageModuleContext{}
+	stageModuleCtx.groupCtx = make([]groupModuleContext, 0, len(plan))
+
+	for _, group := range plan {
+		groupOutcome, newPayload, moduleContexts, reject := executeGroup(executionCtx, group, payload, hookHandler)
+		stageOutcome.ExecutionTimeMillis += groupOutcome.ExecutionTimeMillis
+		stageOutcome.Groups = append(stageOutcome.Groups, groupOutcome)
+		stageModuleCtx.groupCtx = append(stageModuleCtx.groupCtx, moduleContexts)
+		if reject != nil {
+			return stageOutcome, payload, stageModuleCtx, reject
+		}
+
+		payload = newPayload
+	}
+
+	return stageOutcome, payload, stageModuleCtx, nil
+}
+
+func executeGroup[H any, P any](
+	executionCtx executionContext,
+	group hooks.Group[H],
+	payload P,
+	hookHandler hookHandler[H, P],
+) (GroupOutcome, P, groupModuleContext, *RejectError) {
+	var wg sync.WaitGroup
+	done := make(chan struct{})
+	resp := make(chan hookResponse[P])
+
+	for _, hook := range group.Hooks {
+		mCtx := executionCtx.getModuleContext(hook.Module)
+		wg.Add(1)
+		go func(hw hooks.HookWrapper[H], moduleCtx hookstage.ModuleInvocationContext) {
+			defer wg.Done()
+			executeHook(moduleCtx, hw, payload, hookHandler, group.Timeout, resp, done)
+		}(hook, mCtx)
+	}
+
+	go func() {
+		wg.Wait()
+		close(resp)
```

Comment:
Why is the `resp` channel being closed here? This introduces a race condition where `resp` channel would be closed before the `executeHook` have finished sending their results. Better to remove the wait group and add a defer to close the `resp` channel.

Reply:
Perhaps I’m missing something, but in my understanding `wg.Wait()` is a blocking operation, so when `close(resp)` is called, all goroutines have already finished their execution and it’s safe to close the channel.

Reply:
Yes, `wg.Wait()` is blocking and will not unblock until every `executeHook` completes. Let me take another look.

Reply:
Sure, we will add more tests to ensure 100% coverage.

Reply:
We added tests to cover more cases, but right now we cannot achieve 100% code coverage because some of the logic is reachable at later stages only. But we will add more tests in PRs with new stages.